text/0617-cloudfront-oac-l2.md

```diff
@@ -0,0 +1,877 @@
+      originAccessControlConfig: {
+        description: props.description,
+        name: props.originAccessControlName ?? Names.uniqueResourceName(this, {
+          maxLength: 64,
+        }),
+        signingBehavior: props.signing?.behavior ?? SigningBehavior.ALWAYS,
+        signingProtocol: props.signing?.protocol ?? SigningProtocol.SIGV4,
+        originAccessControlOriginType: this.originAccessControlOriginType,
+      },
+    });
+
+    this.originAccessControlId = resource.attrId;
+  }
+
+  private validateOriginAccessControlName(name: string) {
+    if (!Token.isUnresolved(name) && name.length > 64) {
+      throw new Error(`Origin access control name must be 64 characters or less, '${name}' has length ${name.length}`);
+    }
+  }
+}
+```
+
+#### Modifications to `S3BucketOrigin` class
+
+The `S3BucketOrigin` will have two methods, `withAccessIdentity()` and `withAccessControl()`, which each return a class configured with
+the corresponding method of origin access control.
+
+In the case where an imported bucket is being used for the S3 origin, calling `bucket.addToResourcePolicy()` will fail to add the policy statement. Existing
+[workarounds](https://github.com/aws/aws-cdk/issues/6548#issuecomment-869091553) require the user to create a new `BucketPolicy` for the bucket and
+add the policy statements using `bucketPolicy.document.addStatements()`.
+However, this overwrites the whole bucket policy instead of appending statements to the
+existing policy which is a subpar user experience. The proposed solution to this issue is
+to use a custom resource to retrieve the existing bucket policy and append the
+OAC policy statement via the `GetBucketPolicy()` and `PutBucketPolicy()` API calls
+after the CloudFront distribution has been created. Users can choose to opt-in by setting the `overrideImportedBucketPolicy` property to `true`.
+This way we don't silently modify their imported bucket policy which could lead to unintended behaviour.
+
+In the case where the S3 bucket uses SSE-KMS encryption (customer-managed key),
+a circular dependency error occurs when trying to deploy the template. When granting
+the CloudFront distribution access to use the KMS Key, there is a circular dependency:
+
+- CloudFront distribution references the S3 bucket
+- S3 bucket references the KMS key
+- KMS Key references the CloudFront distribution
+
+The proposed solution to this issue is to use a custom resource
+to retrieve and update the KMS key policy after the CloudFront
+distribution has been created via the `GetKeyPolicy()` and `PutKeyPolicy()` API calls.
+
+```ts
+/**
+ * An Origin specific to a S3 bucket (not configured for website hosting).
+ *
+ * Contains additional logic around bucket permissions and origin access control (via OAI or OAC).
+ */
+abstract class S3BucketOrigin extends cloudfront.OriginBase {
+  public static withAccessIdentity(bucket: s3.IBucket, props: S3OriginProps = {}): S3BucketOrigin {
```

Comment:
That being said, I might like this better than what I suggested above for this class. It would be good to add in what this looks like in use by the user (you might have done that below already... I haven't scrolled yet.

Reply:
Oh, what about the option to do neither. Is that a valid option?

Reply:
Yes, it's a valid option to have public origin access (neither OAI nor OAC)